syntax = "proto3";

message ExceptionResponse {
    string module_name = 1;
    string class_name = 2;
    string args = 3;
}

message HealthCheckRequest {
    string client_uri = 1;
    string check_type = 2;
}

message HealthCheckResponse {
    enum Status {
        OK = 0;
        ERROR = 1;
    }
    Status status = 1;
    ExceptionResponse exception = 2;
}

message VersionRequest {}

message VersionResponse {
    string version = 1;
    enum Status {
        OK = 0;
        ERROR = 1;
    }
    Status status = 2;
    ExceptionResponse exception = 3;
}

message ResourceRequest {
    string resource_type = 1;
    string client_uri = 2;
    bytes args = 3;
}

message ResourceResponse {
    string handler = 1;
    enum Status {
        OK = 0;
        ERROR = 1;
    }
    Status status = 2;
    ExceptionResponse exception = 3;
}

message ReleaseResourceRequest {
    string handler = 1;
    string client_uri = 2;
}

message ReleaseResourceResponse {
    enum Status {
        OK = 0;
        ERROR = 1;
    }
    Status status = 2;
    ExceptionResponse exception = 3;
}

message RequestHeader {
    string version = 1;
    string handler = 2;
    string client_uri = 3;
    string method_name = 4;
}

message InstructionRequest {
    oneof instruction {
        RequestHeader header = 1;
        bytes message = 2;
    }
}

message ResponseHeader {
    string version = 1;
    enum Status {
        OK = 0;
        ERROR = 1;
    }
    Status status = 2;
    ExceptionResponse exception = 3;
}

message InstructionResponse {
    oneof instruction {
        ResponseHeader header = 1;
        bytes message = 2;
    }
}

message WriteInstructionsRequest {
    string version = 1;
    string client_uri = 2;
    bytes message = 3;
}

message WriteInstructionsResponse {
    string version = 1;
    enum Status {
        OK = 0;
        ERROR = 1;
    }
    Status status = 2;
    ExceptionResponse exception = 3;
}

service RemoteTrackingService {
    rpc health_check(HealthCheckRequest) returns (HealthCheckResponse) {}
    rpc get_version(VersionRequest) returns (VersionResponse) {}
    rpc get_resource(ResourceRequest) returns (ResourceResponse) {}
    rpc release_resource(ReleaseResourceRequest) returns (ReleaseResourceResponse) {}
    rpc run_instruction(stream InstructionRequest) returns (stream InstructionResponse) {}
    rpc run_write_instructions(stream WriteInstructionsRequest) returns (WriteInstructionsResponse) {}
}
